docs/man/man1/pacparse.1
.TH "pacparse" "1" "" "" ""
.SH "NAME"
pacparse \- tool to parse Proxy Auto\-Config (PAC) files
.SH "SYNOPSIS"
.B pacparse \-p pacfile \-u url [\-h host] [\-c clientip] [\-U pacurl] [\-46edv]
.SH "DESCRIPTION"
.B pacparse 
is a tool to parse Proxy Auto\-Config (PAC) files. It returns the
proxy config string for the given URL and PAC file. 
.B pacparse 
uses the pacparser C library for most of its functionality.
.SH "OPTIONS"
.TP 
.B \-p pacfile
PAC file to parse. Specify "-" to read from standard input.
.TP 
.B \-u url
URL to pass as a parameter to the PAC file's FindProxyForURL function.
.TP 
.B \-h host
Host part of the URL. If not specified, it is determined from the URL.
.TP 
.B \-c clientip
Client's IP address, as returned by the function myIpAddress() in PAC files.
If not specified, it defaults to the IP address associated with the hostname
of the machine on which the tool is running, or 127.0.0.1 if that can't
be found.
.TP 
.B \-U pacurl
URL that the PAC file came from, used to identify the client IP address in
a more reliable way.  The tool parses the host name from the URL,
attempts to connect to each address associated with that host name
with a UDP socket until one is successful, and then uses the IP address
associated with the client side of that socket.
.TP 
.B \-4
Use only IPv4 addresses for -U.
.TP
.B \-6
Use only IPv6 addresses for -U.
.TP 
.B \-e
Enable Microsoft PAC extensions (dnsResolveEx, myIpAddressEx, isResolvableEx).
.TP 
.B \-d
Enable debugging messages.
.TP 
.B \-v
Print version and exit.
.SH "EXAMPLES"
.PP 
To find out the proxy config string for the PAC file "wpad.dat" and the URL
"http://www.google.com":
.PP 
$ pacparse \-p wpad.dat \-u http://www.google.com

For a client with IP address 10.0.12.123:
.PP 
$ pacparse \-p wpad.dat \-c 10.0.12.123 \-u http://www.google.com

For a PAC file hosted at http://wpad/wpad.dat:
.PP 
$ curl \-s http://wpad/wpad.dat | \\
    pacparse \-p \- \-u http://google.com \-U http://wpad/wpad.dat
.SH "SEE ALSO"
pacwget(1),
pacparser_init(3)
.SH "BUGS"
If you have come across a bug in pacparse, please submit a bug report at
https://github.com/pacparser/pacparser/issues.
.SH "AUTHOR"
Written by Manu Garg (http://www.manugarg.com) and Dave Dykstra.
.SH "RESOURCES"
Homepage: https://github.com/pacparser/pacparser.



debian/changelog
pacparser (1.2.6) lucid; urgency=low

  * Feature: Add a feature to record current version in the compiled
    pacparser library and make it available through -v option in pactester.
  * Build: Support for PowerPC. Use va_copy for Linux on all architectures.
    This fixes build for Linux on PowerPC.
  * Build: Fix a major bug because of which pacparser was broken on ARM
    architecture. (http://goo.gl/4ZZ34)
  * Development: Simplify build on windows. Compile spidermonkey from the
    included source (hg logs -r 153:155)

 -- manu garg <manugarg@gmail.com>  Mon, 13 Dec 2010 11:48:33 +0530

pacparser (1.2.1) unstable; urgency=low

  * Implement a subset of Microsoft PAC extensions for IPv6. This adds
    following functions - dnsResolveEx, myIpAddressEx, isResolvableEx. More
    about this change at:
    http://code.google.com/p/pacparser/issues/detail?id=4
  * Use distutils to build python module. This simplifies the build setup a
    lot.
  * Include spidermonkey sources in pacparser tarball (spidermonkey is not
    available as a package in debian anymore) and link spidermonkey library
    statically. This simplifies the distribution.

 -- manu garg <manugarg@gmail.com>  Mon, 09 Aug 2010 23:46:59 +0530

pacparser (1.1.1) unstable; urgency=low

  * Cut a bad release last time. Some changes that didn't go in.

 -- manu garg <manugarg@gmail.com>  Sat, 19 Jun 2010 19:45:33 +0530

pacparser (1.1.0) unstable; urgency=low

  * Match upstream version.

 -- manu garg <manugarg@gmail.com>  Sat, 19 Jun 2010 12:38:35 +0530

pacparser (1.0.9) unstable; urgency=low

  * Match upstream version.
  * Upstream change: remove 'debian' directory from source tarball.

 -- manu garg <manugarg@gmail.com>  Thu, 03 Dec 2009 14:07:24 +0530

pacparser (1.0.8) unstable; urgency=low

  * Big changes to build scripts to make them Mac compatible.
  * Remove spidermonkey source code from pacparser.
  * Add a helper script to download and install spidermonkey.
  * Fix a bug in weekdayRange() utility function.
    (http://code.google.com/p/pacparser/issues/detail?id=2)

 -- manu garg <manugarg@gmail.com>  Tue, 24 Nov 2009 00:02:41 +0530

pacparser (1.0.7) unstable; urgency=low

  * More changes to make it debian friendly.
  * Modified Makefile to produce PIC code for shared library even on i386
    platforms.
  * See http://code.google.com/p/pacparser/source/detail?r=100 for detailed
    list of changes.

 -- manu garg <manugarg@gmail.com>  Sat, 13 Nov 2008 14:48:15 +0530

pacparser (1.0.6-0) unstable; urgency=low

  * Fix some memory leakages in pactester.c (patch by Phil Randal).
  * Make pactester independent of protocol field in URLs.
  * Fix LGPL version in copyright headers.
  * Add copyright headers to pacparser.h.

 -- manu garg <manugarg@gmail.com>  Sat, 10 Oct 2008 14:58:25 +0530

pacparser (1.0.5-0) gutsy; urgency=low

  * Implemented pactester in C using pacparser. Add it to the libpacparser
    package.

 -- manu garg <manugarg@hobbiton.hyd.corp.google.com>  Sat, 10 May 2008 15:23:52 +0530

pacparser (1.0.4-0) gutsy; urgency=low

   * Add a feature to set my ip address (as returned by myIpAddress) to a
     custom value.
   * Fixed the bug reported in Issues (1) due to which myIpAddress function
     was not working on Windows platform.
   * Fixed another bug due to which pacparser failed to parse DOS mode PAC
     files on Windows machines.

 -- manu garg <manugarg@hobbiton.lan>  Fri, 18 Apr 2008 14:27:21 +0530

pacparser (1.0.3-0) gutsy; urgency=low

  * Initial release

 -- manu garg <manugarg@gmail.com>  Mon, 03 Feb 2008 22:08:35 +0530



src/pacparse.c
// This file implements a command called pacparse for downloading and
//  parsing proxy-auto-config files.  Intended to be used from scripts
//  including pacwget.
// Author: Dave Dykstra
//
// This file is based on pactester.c from the pacparser library
// (https://github.com/pacparser/pacparser/blob/master/src/pactester.c)
// Copyright (C) 2008 Manu Garg.
// Author: Manu Garg <manugarg@gmail.com>
//
// pacparse is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// pacparse is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA

#ifndef PACPARSE_VERSION
#define PACPARSE_VERSION "1.0"
#endif

#include "pacparser.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>

int pacdebug=0;
char *progname="pacparse";

void usage()
{
  fprintf(stderr, "\nUsage: %s -p pacfile -u url [-h host] "
          "[-c clientip] [-U pacurl] [-46edv]", progname);
  fprintf(stderr, "\nOptions:\n");
  fprintf(stderr, "  -p pacfile   : PAC file to parse (specify '-' to read "
                  "from standard input)\n");
  fprintf(stderr, "  -u url       : URL parameter to the PAC file's "
  		  "FindProxyForURL function\n");
  fprintf(stderr, "  -h host      : Host part of the URL (default "
  		  "parsed from url)\n");
  fprintf(stderr, "  -c clientip : client IP address (as returned by "
                  "myIpAddress() function\n");
  fprintf(stderr, "                 in PAC files, defaults to IP address "
                  "of client hostname)\n");
  fprintf(stderr, "  -U pacurl    : URL that pacfile came from, to identify "
  		   "client IP address\n");
  fprintf(stderr, "                 (creates UDP socket to host, more reliable "
  		  "than default\n");
  fprintf(stderr, "                  on clients with multiple IP addresses)\n");
  fprintf(stderr, "  -4           : use only IPv4 addresses for -U\n");
  fprintf(stderr, "  -6           : use only IPv6 addresses for -U\n");
  fprintf(stderr, "  -e           : enable microsoft extensions "
                  "(Ex functions)\n");
  fprintf(stderr, "  -d           : enable debugging messages\n");
  fprintf(stderr, "  -v           : print version and exit\n");
}

char *get_host_from_url(const char *url)
{
  // copy  url to a  pointer that we'll use to seek through the string.
  char *p = strdup(url);
  // Move to :
  while (*p != ':' && *p != '\0')
    p++;
  if (p[0] == '\0'||                    // We reached end without hitting :
      p[1] != '/' || p[2] != '/'        // Next two characters are not //
      ) {
    fprintf(stderr, "%s: Not a proper URL\n", progname);
    return NULL;
  }
  p = p + 3;                            // Get past '://'
  // Host part starts from here.
  char *host = p;
  if (*p == '\0' || *p == '/' || *p == ':') {   // If host part is null.
    fprintf(stderr, "%s: Not a proper URL\n", progname);
    return NULL;
  }
  // Seek until next /, : or end of string.
  while (*p != '/' && *p != ':' && *p != '\0')
    p++;
  *p = '\0';
  return host;
}

void set_myip_from_host(const char *host, int ipversion)
{
  int err, fd;
  struct addrinfo addrinfo, *res, *r;
  struct sockaddr_in sinbuf;
  struct sockaddr_in6 sin6buf;
  socklen_t namelen;
  char errbuf[4096];
  char ipbuf[256];

  memset(&addrinfo, 0, sizeof(addrinfo));
  switch(ipversion)
  {
    case 4:
      addrinfo.ai_family = AF_INET;
      break;
    case 6:
      addrinfo.ai_family = AF_INET6;
      break;
    default:
      addrinfo.ai_family = AF_UNSPEC;
      break;
  }
  addrinfo.ai_socktype = SOCK_DGRAM;
  addrinfo.ai_protocol = IPPROTO_UDP;

  if ((err = getaddrinfo(host, 0, &addrinfo, &res)) != 0) {
    fprintf(stderr, "%s: error from getaddrinfo on %s: %s\n", 
    		progname, host, gai_strerror(err));
    exit(2);
  }
  errbuf[0] = '\0';
  ipbuf[0] = '\0';
  for (r = res; r; r = r->ai_next) {
    if (pacdebug && (errbuf[0] != '\0')) {
      fprintf(stderr, "DEBUG: trying next addr after %s\n", errbuf);
    }
    if ((fd = socket(r->ai_family, r->ai_socktype, r->ai_protocol)) < 0) {
      fprintf(stderr, "%s: error creating socket for %s: %s\n",
		  progname, host, strerror(errno));
      exit(2);
    }
    if (connect(fd, r->ai_addr, r->ai_addrlen) < 0) {
      snprintf(errbuf, sizeof(errbuf),
		  "error connecting UDP socket(s) to %s, last error: %s",
		      host, strerror(errno));
      close(fd);
      continue;
    }
    if (r->ai_family == AF_INET) {
      namelen = sizeof(sinbuf);
      if (getsockname(fd, (struct sockaddr *)&sinbuf, &namelen) < 0) {
	fprintf(stderr, "%s: error on getsockname from socket to %s: %s\n",
		  progname, host, strerror(errno));
	exit(2);
      }
      if (inet_ntop(AF_INET, &sinbuf.sin_addr, ipbuf, sizeof(ipbuf)) == 0) {
	fprintf(stderr, "%s: error on inet_ntop from getsockname to %s: %s\n",
		  progname, host, strerror(errno));
	exit(2);
      }
    }
    else if (r->ai_family == AF_INET6) {
      namelen = sizeof(sin6buf);
      if (getsockname(fd, (struct sockaddr *)&sin6buf, &namelen) < 0) {
	fprintf(stderr, "%s: error on getsockname from socket to %s: %s\n",
		  progname, host, strerror(errno));
	exit(2);
      }
      if (inet_ntop(AF_INET6, &sin6buf.sin6_addr, ipbuf, sizeof(ipbuf)) == 0) {
	fprintf(stderr, "%s: error on inet_ntop from getsockname to %s: %s\n",
		  progname, host, strerror(errno));
	exit(2);
      }
    }
    else {
      fprintf(stderr, "%s: unknown address family %d\n", 
	progname, r->ai_family);
      exit(2);
    }
    break;
  }
  freeaddrinfo(res);
  close(fd);
  if (ipbuf[0] == '\0') {
    if (errbuf[0] == '\0')
      fprintf(stderr, "%s: could not determine IP for %s, error unknown\n",
      	progname, host);
    else
      fprintf(stderr, "%s: %s\n", progname, errbuf);
    exit(2);
  }
  if (pacdebug) {
    fprintf(stderr, "DEBUG: Setting myip to %s\n", ipbuf);
  }
  pacparser_setmyip(ipbuf);
}

int main(int argc, char* argv[])
{
  char *pacfile=NULL, *url=NULL, *host=NULL, *clientip=NULL, *pacurl=NULL;
  int ipversion = 0;
  int enable_microsoft_extensions = 0;
  signed char c;
  while ((c = getopt(argc, argv, "edv46p:u:h:c:U:")) != -1)
    switch (c)
    {
      case 'v':
        printf("%s %s; pacparser library %s\n", progname, PACPARSE_VERSION,
				pacparser_version());
        return 0;
      case 'p':
        pacfile = optarg;
        break;
      case 'u':
        url = optarg;
        break;
      case 'h':
        host = optarg;
        break;
      case 'c':
        clientip = optarg;
        break;
      case 'U':
        pacurl = optarg;
        break;
      case '4':
	ipversion = 4;
	break;
      case '6':
        ipversion = 6;
	break;
      case 'e':
        enable_microsoft_extensions = 1;
        break;
      case 'd':
        pacdebug = 1;
	putenv("PACPARSER_DEBUG=1");
	putenv("DEBUG=1"); /* for older versions of the library */
        break;
      case '?':
        usage();
        return 1;
      default:
        abort ();
    }

  if (!pacfile) {
    fprintf(stderr, "%s: You didn't specify the PAC file\n", progname);
    usage();
    return 1;
  }

  if(enable_microsoft_extensions)
    pacparser_enable_microsoft_extensions();

  // initialize pacparser
  if (!pacparser_init()) {
      fprintf(stderr, "%s: Could not initialize pacparser\n", progname);
      return 1;
  }

  // Read pacfile from stdin
  if (strcmp("-", pacfile) == 0) {
    char *script;
    int buffsize = 4096;
    int maxsize = 1024 * 1024;          // Limit the max script size to 1 MB
    size_t script_size = 1;             // For the null terminator
    char buffer[buffsize];

    script = (char*) malloc(sizeof(char) * buffsize);
    if (script == NULL) {
      fprintf(stderr,"%s: Failed to allocate memory for the script\n",
      	progname);
      return(1);
    }
    script[0] = '\0';                   // Null terminate to prepare for strcat

    while(fgets(buffer, buffsize, stdin)) {
      if (strlen(buffer) == 0) break;
      char *old = script;
      script_size += strlen(buffer);
      if (script_size > maxsize) {
        fprintf(stderr, "Input file is too big. Maximum allowed size is: %d",
                maxsize);
        free(script);
        return 1;
      }
      script = realloc(script, script_size);
      if (script == NULL) {
        fprintf(stderr, "%s: Failed to realloc %d bytes of memory for the script %s\n",
	  progname, (int)script_size, script);
        free(old);
        return 1;
      }
      strcat(script, buffer);
    }

    if (ferror(stdin)) {
      free(script);
      fprintf(stderr, "%s: Error reading from stdin\n", progname);
      return 1;
    }

    if(!pacparser_parse_pac_string(script)) {
      fprintf(stderr, "%s: Could not parse the pac script:\n%s\n",
              progname, script);
      free(script);
      pacparser_cleanup();
      return 1;
    }
    free(script);
  }
  else {
    if(!pacparser_parse_pac_file(pacfile)) {
      fprintf(stderr, "%s: Could not parse the pac file: %s\n",
              progname, pacfile);
      pacparser_cleanup();
      return 1;
    }
  }

  if(clientip)
    pacparser_setmyip(clientip);
  else if(pacurl) {
    char *pachost;
    pachost = get_host_from_url(pacurl);
    if (!pachost) {
      fprintf(stderr, "%s: Error finding hostname in %s\n", progname, pacurl);
      exit(2);
    }
    set_myip_from_host(pachost, ipversion);
  }

  char *proxy;

  if (url) {
    if (!host)
      host = get_host_from_url(url);
    if (host) {
      proxy = NULL;
      proxy = pacparser_find_proxy(url, host);
      if (proxy == NULL) {
        fprintf(stderr, "%s: Problem in finding proxy for %s\n", progname, url);
        pacparser_cleanup();
        return 1;
      }
      if (proxy) printf("%s\n", proxy);
    }
  }

  pacparser_cleanup();
  return 0;
}
